### DISCUSSION TITLE
Sandboxing Opal code

### DISCUSSION THREAD

**Discussion by MarketingPip** [opal/opal#2488] [Q&A]
Sorry if I am missing this blindly in-front of my eyes, but is there anyway I can disable usage of JavaScript functions on the DOM? Trying to build a client side interpreter with this but don't want to allow JS obviously inputted by user in Ruby.

As well is there a default over-ride command etc that I can force all outputs such as ```put``` to a DOM / HTML element? 

Thank you to anybody that answers this for me in advanced.

**Comment by hmdne** [comment]
What I understand here is that you want to sandbox the Opal code and it's possible and not even that hard :D (but - you need to use facilities provided by your web browser - namely WebWorker, and possibly a couple more, like using iframes ( https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe - see the sandbox attribute ) or serving it from a different domain with carefully crafted `Content-Security-Policy` headers).

In short, the Opal-compiled code can run in a webworker and you can use the CDN releases for that, or you can compile Opal (possibly with the opal-parser library).

I have a small example of a program that runs Opal in a WebWorker, please take a look at this file:

https://github.com/ruby/TryRuby/pull/141/commits/2967d90feebe695e3481be55598b18af41adc691#diff-daa4fd06754d5fa22a26f4bf9f211742ac0a2987072beb756467217ba31c5fdc

What we have here is a little bit more complicated - it's an Opal program that runs Opal (from our CDN) in a WebWorker (and interfaces with it), but a lot of logic described here is with simple JavaScript - do note though, that this doesn't deal with sandboxing (the code had different goals) - while user can't manipulate the DOM, he may be able to send network requests from a context of the main domain, possibly creating a situation where XSRF and cookie stealing is possible.

Another tidbit for understanding the code I posted is that we use an `autoawait` facility, which makes it so that all the functions that have names ending with `_await` are executed with an `await` JS keyword - see: https://opalrb.com/docs/guides/v1.6.0/async .

For your other question about redirecting `stdout` and `stderr`: the logic of what our $stdout and $stderr do is described with a #write_proc property set on them. Please take a look at the default implementation we have: https://github.com/opal/opal/blob/master/opal/corelib/io.rb#L308 (the platforms, like let's say, the GJS platform does override those procs: https://github.com/opal/opal/blob/master/stdlib/gjs/io.rb#L16 - but also the code I linked does something similar in a `prepare_stdio_await` method)

**Reply by hmdne** [reply]
As I sidenote, I think my patch to the TryRuby program is actually secure, because the WebWorker runs from a `data:` URI which according to https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy creates a fresh security context. It will still be able to execute network requests, but shouldn't pose the security issue to your domain at all.

**Comment by MarketingPip** [comment]
@hmdne I appreciate the in-depth details & fast response. I will be sand boxing my interpreter inside of a iframe. And to be honest - I do not have experience with Ruby so I was hoping I could possibly get passed a simple pure JS function to compile Ruby (without any JS / JS API features from Opal inside of the sandbox) so it doesn't confuse users etc. Not that I meaning to dismiss all of the above - just was hoping maybe their was a function I could call that allowed code evaluation with option to disable JS API - in the most newb / spoon fed way - ie a basic Try / Catch function with evalRuby etc... (I was hoping for :+1: )

**Comment by hmdne** [comment]
@MarketingPip I have rewritten the code referenced above in pure JavaScript for you (below).

>that allowed code evaluation with option to disable JS API

This is unfortunately out of current scope of Opal. We reuse the backticks (\`) which in Ruby mean "execute in shell, eg. bash, and return stdout" to mean "compile as JavaScript verbatim". There is a special method `JS` which allows for calling a JS function or getting/setting a JS property (eg. ``` `document`.JS['location'].JS.replace('https://opalrb.com/') ```), but there may be other places where escape to JS is possible.

It may be possible to disable those, but no public API for that exists in the current version.

Opal itself, though, doesn't include browser APIs, like DOM manipulation - this is handled by the Opal-Browser project. Opal itself is meant to run on various JS runtimes, not necessarily browser, but also Node, Deno (in the next release), GJS, WebWorker (which supports a limited set of browser APIs itself).

Now, depending on which usecase you want:

- If you want it to be a verbatim representation of Ruby, for example as a part of an interactive course, be aware that Opal is an alternative Ruby implementation and may not have every Ruby feature out there and there are a couple of incompatibilities - see: https://opalrb.com/docs/guides/v1.6.0/unsupported_features . The "exit to JS" features should be obscure enough for people not to notice.
  - Ruby is planning on supporting being compiled to WASM in the 3.2 release (to be released on 26th of December). If full and exact Ruby compatibility is a goal for you, then perhaps this can be a good choice to consider: https://github.com/ruby/ruby.wasm (do note that unfortunately this will be a much heavier download)

- If you plan on allowing users to upload their own code to be ran by other users and security is of your utmost importance, "exit to JS" in a WebWorker is not a problem, provided that (as mentioned above) you do proper sandboxing. As I mentioned in the follow-up comment, running user-provided code in a WebWorker should be enough (as long as you keep loading the "kernel" from the data: URI which resets the security context). User would still be able to trigger network connections to other websites - in the same way that visiting a random page would (eg. no risk of cookie stealing). If even that's too much, you will need to load the WebWorker "kernel" - best from a separate domain - where you set the Content-Security-Policy header according to your needs: https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP ; https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy

As a separate note, this loads Opal from the CDN. You may as well choose to compile Opal yourself (`gem install opal; opal -ropal-parser -c -e '#' > opal-and-opal-parser.js`) if you want more flexibility, even embed it inside the "kernel" to improve evaluation times, furthermore even compile user-provided applications server-side and serve just JavaScript (do note that Opal has not been audited for that case, to be safe you should sandbox the compiler process).

```html
<!doctype html>
<html>
<head>
<title>Sample Opal WebWorker Application</title>
</head>
<body>
<h1>Sample Opal WebWorker Application</h1>

<script>
"use strict";

class WorkerWrap {
    constructor(url) {
        this.worker = new Worker(url);
    }
    
    send(message) {
        this.worker.postMessage(message);
    }
    
    onMessage(block) {
        this.worker.onmessage = block;
    }
    
    close() {
        this.worker.terminate();
    }
}

const AWAIT_LOGIC_KERNEL = `
    onmessage = function(e) {
        var result;
        try {
            console.log("<<", e.data);
            switch (e.data[0]) {
                case "script":
                    result = importScripts(e.data[1]);
                    break;
                case "eval":
                    result = eval(e.data[1]);
                    break;
                case "nop":
                    break;
                default:
                    result = globalThis[e.data[0]](e.data[1]);
            }
            if(e.data[2]) postMessage(["ready", result]);
        }
        catch(err) {
            if(e.data[2]) postMessage(["ready", result]);
            if(globalThis.handleError) globalThis.handleError(err);
        }
    }
`;

class AwaitLogicWorker extends WorkerWrap { 
    constructor() {
        super("data:application/javascript;base64,"+btoa(AWAIT_LOGIC_KERNEL));
        this.onMessage(this.handleMessage.bind(this));
    }
    
    async sendAwait(msg, arg) {
        var self = this;
        this.waiter = new Promise(function(resolve, reject) {
            self.waiterResolve = resolve;
        });
        this.send([msg, arg, true])
        return await this.waiter;
    }

    handleMessage(msg) {
        console.log(">>", msg.data);

        var arg = msg.data[1];
        var msg = msg.data[0];
        switch(msg) {
            case "ready":
                this.waiterResolve(arg);
                this.waiter = this.waiterResolve = null;
                break;
            default:
                this["handle_"+msg](arg);
        }
    }
}

class OpalWorker extends AwaitLogicWorker {
    constructor(version) {
        super();
        this.version = version;
    }
    
    async uploadOpalAwait() {
        await this.uploadScriptAwait("opal");
        await this.uploadScriptAwait("opal-parser");
    }
    
    async uploadScriptAwait(name) {
        await this.sendAwait("script", "https://cdn.opalrb.com/opal/"+this.version+"/"+name+".min.js");
        if (name != "opal") await this.sendAwait("eval", "Opal.require("+JSON.stringify(name)+"),null");
    }
    
    async evalRubyAwait(ruby) {
        return await this.sendAwait("eval", "var out = eval(Opal.compile("+JSON.stringify(ruby)+")); out === Opal.nil ? '' : out.$inspect();");
    }
    
    async prepareStdioAwait(writer) {
        this.writer = writer
        await this.evalRubyAwait(`
            $stdout.write_proc = $stderr.write_proc = ->(str) do
                \`postMessage(["stdout", str])\`
            end
        `);
    }
    
    handle_stdout(str) {
        this.writer("stdout", str);
    }

    async prepareErrorHandlerAwait() {
        await this.evalRubyAwait(`
            error_handler = proc do |err|
                backtrace = err.backtrace
                backtrace = backtrace.select { |i| i.include? '<anonymous>' }
                backtrace = backtrace.map { |i| i.gsub(/.*(<anonymous>)/, '\\1') }
                backtrace = ["(file)"] if backtrace.empty?
                err.set_backtrace(backtrace)
                err = err.full_message
                \`postMessage(["error", err])\`
            end
            \`globalThis.handleError = error_handler\`
        `);
    }

    handle_error(err) {
        this.writer("error", err);
    }
}
</script>

<form action='#' id='rubyform'>
    <input type='text' id='input'>
    <input type='submit'>
    
    <pre id='output'></pre>
</form>

<script>
"use strict";

var input = document.getElementById("input");
var output = document.getElementById("output");

async function runRubyCode(e) {
    e.preventDefault();
    
    output.replaceChildren();
    
    var opalWorker = new OpalWorker("1.6.0");
    await opalWorker.uploadOpalAwait();
    await opalWorker.prepareErrorHandlerAwait();
    await opalWorker.prepareStdioAwait(function(type, ret) {
        var span = document.createElement("SPAN");
        if (type == 'error') span.style.color = "red";
        span.innerText = ret;
        output.appendChild(span);
    });
    
    await opalWorker.evalRubyAwait(input.value);
    
    // Ensure the errors arrive
    await opalWorker.sendAwait("nop");
    
    opalWorker.close();
}

document.getElementById("rubyform").addEventListener("submit", runRubyCode);
</script>
</body>
</html>
```

**Reply by MarketingPip** [reply]
@hmdne you have no idea how much I appreciate your time and effort spent on that reply. Thank you VERY much.


I was under the impression Opal allowed for DOM manipulation by default etc. Similar to Brython (python in the browser - which allows users to use a Python Import - to access the DOM) - which I do NOT want as said. 👌 

So thank you for clarifying everything for me as well again thank you so much for literally SPOON FEEDING everything I needed. 👌👍 (we need more superman like you in the world) hahah! 🔥
